Fast path atomic write batches on one phase commit txns

spencerkimball opened the pull request:
Added new 1PC stat to the txn metrics collected by the txn coordinator.
Fixed a bug in the transaction coordinator which wasn't counting txn
stats in the event of 1PC transactions.

Removed unused resolved intents from `EndTransactionResponse`.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5966)

<!-- Reviewable:end -->

spencerkimball committed:
Fast path atomic write batches on one phase commit txns

Added new 1PC stat to the txn metrics collected by the txn coordinator.
Fixed a bug in the transaction coordinator which wasn't counting txn
stats in the event of 1PC transactions.

Removed unused resolved intents from `EndTransactionResponse` and added
a boolean indicating whether the txn was done with a single phase commit.

spencerkimball commented:
```
name                    old time/op  new time/op  delta
KVInsert1_Native-4       844µs ± 1%   439µs ± 1%  -48.01%    (p=0.000 n=9+9)
KVInsert1_SQL-4          660µs ± 1%   587µs ± 1%  -11.00%  (p=0.000 n=10+10)
KVInsert10_Native-4     1.11ms ± 1%  0.64ms ± 1%  -42.61%  (p=0.000 n=10+10)
KVInsert10_SQL-4        1.33ms ± 1%  1.09ms ± 1%  -18.30%  (p=0.000 n=10+10)
KVInsert100_Native-4    3.40ms ± 1%  2.40ms ± 2%  -29.53%  (p=0.000 n=10+10)
KVInsert100_SQL-4       7.33ms ± 2%  5.46ms ± 3%  -25.48%  (p=0.000 n=10+10)
KVInsert1000_Native-4   27.2ms ± 2%  20.3ms ± 2%  -25.38%  (p=0.000 n=10+10)
KVInsert1000_SQL-4      75.9ms ± 1%  55.3ms ± 4%  -27.10%  (p=0.000 n=10+10)
KVInsert10000_Native-4   332ms ± 7%   260ms ± 5%  -21.80%  (p=0.000 n=10+10)
KVInsert10000_SQL-4      1.06s ± 6%   0.72s ± 4%  -32.40%  (p=0.000 n=10+10)
KVUpdate1_Native-4      1.17ms ± 1%  0.74ms ± 0%  -36.45%   (p=0.000 n=10+9)
KVUpdate1_SQL-4         1.05ms ± 1%  0.97ms ± 2%   -7.39%   (p=0.000 n=9+10)
KVUpdate10_Native-4     1.65ms ± 1%  1.17ms ± 1%  -29.35%  (p=0.000 n=10+10)
KVUpdate10_SQL-4        1.72ms ± 2%  1.54ms ± 1%  -10.53%  (p=0.000 n=10+10)
KVUpdate100_Native-4    5.78ms ± 2%  4.35ms ± 1%  -24.69%  (p=0.000 n=10+10)
KVUpdate100_SQL-4       7.30ms ± 3%  5.91ms ± 1%  -19.05%  (p=0.000 n=10+10)
KVUpdate1000_Native-4   44.9ms ± 2%  34.1ms ± 1%  -23.98%   (p=0.000 n=10+9)
KVUpdate1000_SQL-4      62.0ms ± 2%  47.1ms ± 1%  -24.02%  (p=0.000 n=10+10)
KVUpdate10000_Native-4   498ms ± 3%   424ms ± 3%  -14.85%  (p=0.000 n=10+10)
KVUpdate10000_SQL-4      775ms ± 2%   653ms ± 2%  -15.81%  (p=0.000 n=10+10)
KVDelete1_Native-4       857µs ± 1%   444µs ± 1%  -48.17%   (p=0.000 n=9+10)
KVDelete1_SQL-4          690µs ± 2%   628µs ± 2%   -9.01%  (p=0.000 n=10+10)
KVDelete10_Native-4     1.18ms ± 1%  0.67ms ± 1%  -43.03%  (p=0.000 n=10+10)
KVDelete10_SQL-4        1.08ms ± 2%  0.96ms ± 2%  -11.03%  (p=0.000 n=10+10)
KVDelete100_Native-4    3.75ms ± 3%  2.42ms ± 1%  -35.49%  (p=0.000 n=10+10)
KVDelete100_SQL-4       4.19ms ± 2%  3.62ms ± 1%  -13.57%  (p=0.000 n=10+10)
KVDelete1000_Native-4   30.2ms ± 1%  19.1ms ± 1%  -36.76%  (p=0.000 n=10+10)
KVDelete1000_SQL-4      36.0ms ± 2%  29.7ms ± 2%  -17.45%  (p=0.000 n=10+10)
KVDelete10000_Native-4   305ms ± 2%   233ms ± 1%  -23.36%   (p=0.000 n=10+9)
KVDelete10000_SQL-4      610ms ± 8%   404ms ± 5%  -33.85%   (p=0.000 n=10+9)
KVScan1_Native-4         290µs ± 1%   248µs ± 2%  -14.72%    (p=0.000 n=9+9)
KVScan1_SQL-4            428µs ± 2%   430µs ± 2%     ~     (p=0.529 n=10+10)
KVScan10_Native-4        333µs ± 1%   289µs ± 1%  -13.16%  (p=0.000 n=10+10)
KVScan10_SQL-4           526µs ± 1%   523µs ± 1%     ~     (p=0.052 n=10+10)
KVScan100_Native-4       617µs ± 2%   569µs ± 3%   -7.84%  (p=0.000 n=10+10)
KVScan100_SQL-4         1.35ms ± 2%  1.33ms ± 2%     ~      (p=0.113 n=9+10)
KVScan1000_Native-4     3.16ms ± 1%  2.99ms ± 2%   -5.65%    (p=0.000 n=9+9)
KVScan1000_SQL-4        8.85ms ± 1%  8.49ms ± 1%   -4.08%   (p=0.000 n=10+9)
KVScan10000_Native-4    32.5ms ± 1%  31.1ms ± 2%   -4.33%   (p=0.000 n=8+10)
KVScan10000_SQL-4       93.1ms ± 4%  90.1ms ± 4%   -3.15%  (p=0.002 n=10+10)
```

petermattis commented:
Those numbers are compelling. @tschottdorf will have to chime in about the `storage` changes.

---

Review status: 0 of 20 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[roachpb/batch.go, line 89 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4W_WYFo9wqsUO1S4a:-KF4W_WYFo9wqsUO1S4b:389548781) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/roachpb/batch.go#L89)):*
Feels odd to me to use `Method() != BeginTransaction` here when `GetInner()` is an interface. While it is more code, I think this would would be better as:

```
if _, ok := ba.Requests[0].GetInner().(*BeginTransactionRequest); !ok {
  return false
}
if _, ok := ba.Requests[len(ba.Requests)-1].GetInner().(*EndTransactionRequest); !ok [
  return false
}
```

---

*[roachpb/data.proto, line 289 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4X9ZA3x4FkisIFGwO:-KF4X9ZB7vgSj0ifj2sQ:1171171877) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/roachpb/data.proto#L289)):*
This changes the encoding of `Transactions`, even if `one_phase_commit` is not set. Is that problematic?

This field seems like it should be part of `EndTransactionResponse`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-:-KF4YV7l3n07Elq1lNjD:-1463787230)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 20 of 20 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[kv/txn_coord_sender.go, line 852 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4UvlYkpp44toT8XJM:-KF4UvlYkpp44toT8XJN:-1142835268) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/kv/txn_coord_sender.go#L852)):*
isn't this just ` } else {`?

---

*[kv/txn_coord_sender_test.go, line 986 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4V_WmV-sVgDOyIyn0:-KF4V_WmV-sVgDOyIyn1:288656012) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/kv/txn_coord_sender_test.go#L986)):*
Should document (somewhere) that a `1PC` transaction which experiences restarts is not counted here.

---

*[kv/txn_coord_sender_test.go, line 1067 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4VOqj4jsWZS6mYVeZ:-KF4VOqj4jsWZS6mYVe_:542710335) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/kv/txn_coord_sender_test.go#L1067)):*
s/\./ works./

---

*[roachpb/api.proto, line 240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4VjqPCAxBJS5U2bzZ:-KF4VjqPCAxBJS5U2bz_:397660837) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ac77075268a6957196217197742f6f8e98556fc/roachpb/api.proto#L240)):*
Was this field populated before? If so, unfortunately a migration is necessary.

---

*[roachpb/batch.go, line 78 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4WAVRBBIepONLOJTg:-KF4WAVRBBIepONLOJTh:1845486789) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/roachpb/batch.go#L78)):*
Move this to `replica.go`. Being 1PC isn't a property of the batch until it's past the points where it's split up.

---

*[roachpb/batch.go, line 81 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4WTcH93X9E9lk-8IH:-KF4WTcH93X9E9lk-8II:2040701482) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/roachpb/batch.go#L81)):*
the checks in this method are not obvious. I would have expected

```go
if _, hasBegin := ba.GetArg(BeginTransaction); !hasBegin {
  return false
}
if _, hasEnd := ba.GetArg(EndTransaction); !hasEnd {
  return false
}
return true
```

Mind using that or adding comments to the code you have there. In the former case, might make sense to adjust `GetArg` to only look at the first entry when searching for `BeginTransaction`.

---

*[roachpb/data.proto, line 289 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4XSNuAJXwx2a59-RG:-KF4XSNuAJXwx2a59-RH:195610557) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/roachpb/data.proto#L289)):*
Does this need to be on the proto? Isn't the sequence number the correct indicator already? `OnePhaseCommit == (Sequence == 1) && (Status == Committed)`. Or, if anything, it should be on `EndTransactionResponse` (I suppose you need it somewhere for testing).

---

*[storage/replica.go, line 1690 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4YC0Xuky81RFLzFAa:-KF4YC0Yf8NcrULY2mBo:-255367742) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/storage/replica.go#L1690)):*
s/stripped/strippedBA/

---

*[storage/replica.go, line 1720 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4Yjn65QY7azVlvDHN:-KF4Yjn8l_QTAc-4XPCN:-772567131) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/storage/replica.go#L1720)):*
Instead of stripping `{Begin,End}Transaction`, consider replacing them with `NoopRequest`. That way, you'll get a `br.Responses` of the right length back - might look/feel better.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[kv/txn_coord_sender_test.go, line 986 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4V_WmV-sVgDOyIyn0:-KF5HUE5OK8_AXS-wLKO:1887296068) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/kv/txn_coord_sender_test.go#L986)):*
I don't understand the comment. Could you elaborate?

---

*[roachpb/api.proto, line 240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4VjqPCAxBJS5U2bzZ:-KF5HhcJorh0lKDfLX4w:510826718) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ac77075268a6957196217197742f6f8e98556fc/roachpb/api.proto#L240)):*
Never. It was never used at all.

---

*[roachpb/data.proto, line 289 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4X9ZA3x4FkisIFGwO:-KF5MSNighXMlEKphg1o:320426228) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/roachpb/data.proto#L289)):*
OK, removed it from `Transaction` and added it to `EndTransactionResponse`. arguably we don't even need it there, but I'd rather have it for clarity.

---

*[roachpb/data.proto, line 289 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4XSNuAJXwx2a59-RG:-KF5Nu8D9UyrUa6PBZMf:5174721) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/roachpb/data.proto#L289)):*
pl

---

*[storage/replica.go, line 1720 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4Yjn65QY7azVlvDHN:-KF5O7xXjMze-Wtjq5qQ:-785192780) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/storage/replica.go#L1720)):*
Well I am now responding with an `EndTransactionResponse` that has `OnePhaseCommit` set to `true`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966)*
<!-- Sent from Reviewable.io -->

petermattis commented on roachpb/api.proto:
```diff
@@ -236,8 +236,10 @@ message EndTransactionResponse {
-  // List of intents resolved by EndTransaction call.
-  repeated bytes resolved = 3 [(gogoproto.casttype) = "Key"];
+  // Obsolete tag. Removed in #5966.
+  repeated bytes obsolete_resolved = 3;
```
(Reviewable is being annoying, so back to github)

We don't have a style for obsolete or deprecate tags, but I'd prefer something that is louder: `OBSOLETE_removed` or `DEPRECATED_removed`.

spencerkimball replied:
Changed to `DEPRECATED`.

petermattis commented:
LGTM, though wait for @tschottdorf's sign-off as well.

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with some comments.

---

Reviewed 9 of 14 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[kv/txn_coord_sender_test.go, line 986 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4V_WmV-sVgDOyIyn0:-KF65YHcMQ4cayLxWqXS:-258078999) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/kv/txn_coord_sender_test.go#L986)):*
Looking at this counter it wasn't immediately obvious to me what it counts. Does it count transactions which successfully commit in one total round trip from the coordinator? What about one that would be a 1PC txn except that it catches a `WriteWithinUncertaintyIntervalError`? That would not be counted. Specifying what a 1PC commit somewhere can't hurt. It's a minor point.

---

*[roachpb/api.proto, line 240 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KEy5hwBcNf8KAMkmGlb-r2-240:-KF65tR18-bNqTH2OOvR:318682851) ([raw file](https://github.com/cockroachdb/cockroach/blob/b206c4c846132b9623cd11fd0cf056763eef5ada/roachpb/api.proto#L240)):*
Not changed yet (but probably publish-before-push). Just make sure it's there before you merge.

---

*[roachpb/api.proto, line 241 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF68GzAfsYyAEe4yG8_:-KF68GzAfsYyAEe4yG8a:-877618261) ([raw file](https://github.com/cockroachdb/cockroach/blob/b206c4c846132b9623cd11fd0cf056763eef5ada/roachpb/api.proto#L241)):*
Comment could explain what that means.

---

*[storage/replica.go, line 1720 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4Yjn65QY7azVlvDHN:-KF669i9nl48-AyIE98u:-314420266) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/storage/replica.go#L1720)):*
My suggestion is to change the definition of `strippedBA`: mask out the first and last request with a `NoopRequest{}` (can be a singleton). That way, you get a response slice back that has the right length.

---

*[storage/replica.go, line 1694 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF67QT3jW-p6qjVdeXI:-KF67QT4OICt2XMbZNRz:-30012406) ([raw file](https://github.com/cockroachdb/cockroach/blob/b206c4c846132b9623cd11fd0cf056763eef5ada/storage/replica.go#L1694)):*
What errors which can go away with the transaction do we expect here? ReadWithinUncertaintyIntervalError? Looks that we don't need to (should not) execute a second time on most errors here. Can we white-list them?
Ok as a TODO, but maybe it's trivial.

---

*[storage/replica.go, line 1737 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF68MV0zv2Mc3hDbPXa:-KF68MV0zv2Mc3hDbPXb:-638431424) ([raw file](https://github.com/cockroachdb/cockroach/blob/b206c4c846132b9623cd11fd0cf056763eef5ada/storage/replica.go#L1737)):*
Is this expected in practice?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-:-KF686ktxoePFfDb2pC6:-913190055)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
@spencerkimball, I ran some of the standard SQL benchmarks against this change, and the results look great (especially insert and delete)!

```
name                                old time/op    new time/op    delta
Select1_Cockroach-2                   45.8µs ± 1%    46.2µs ± 1%   +0.88%          (p=0.003 n=9+9)
Select2_Cockroach-2                   1.27ms ± 3%    1.28ms ± 6%     ~           (p=0.912 n=10+10)
Insert1_Cockroach-2                    753µs ± 1%     681µs ± 1%   -9.53%        (p=0.000 n=10+10)
Insert10_Cockroach-2                  1.13ms ± 1%    0.97ms ± 1%  -13.77%         (p=0.000 n=9+10)
Insert100_Cockroach-2                 4.86ms ± 2%    3.59ms ± 1%  -26.21%         (p=0.000 n=10+9)
Insert1000_Cockroach-2                46.4ms ± 3%    31.9ms ± 6%  -31.18%         (p=0.000 n=9+10)
Update1_Cockroach-2                   1.24ms ± 1%    1.16ms ± 2%   -6.26%         (p=0.000 n=9+10)
Update10_Cockroach-2                  2.09ms ± 1%    1.88ms ± 2%   -9.72%          (p=0.000 n=9+9)
Update100_Cockroach-2                 9.66ms ± 4%    8.02ms ± 4%  -17.05%        (p=0.000 n=10+10)
Update1000_Cockroach-2                82.2ms ± 4%    65.6ms ± 4%  -20.27%        (p=0.000 n=10+10)
Delete1_Cockroach-2                   1.03ms ± 4%    0.87ms ± 4%  -14.94%        (p=0.000 n=10+10)
Delete10_Cockroach-2                  1.98ms ± 2%    1.17ms ± 1%  -40.85%        (p=0.000 n=10+10)
Delete100_Cockroach-2                 13.0ms ± 4%     5.3ms ± 9%  -59.37%         (p=0.000 n=9+10)
Delete1000_Cockroach-2                 130ms ± 8%      47ms ± 6%  -63.77%          (p=0.000 n=9+9)
Scan1_Cockroach-2                      477µs ± 2%     471µs ± 2%   -1.28%         (p=0.033 n=10+7)
Scan10_Cockroach-2                     535µs ± 1%     538µs ± 1%   +0.56%         (p=0.004 n=10+9)
Scan100_Cockroach-2                   1.01ms ± 2%    1.00ms ± 3%     ~           (p=0.280 n=10+10)
Scan1000_Cockroach-2                  5.48ms ± 2%    5.38ms ± 2%   -1.83%        (p=0.002 n=10+10)
Scan10000_Cockroach-2                 60.6ms ± 3%    59.2ms ± 3%   -2.39%        (p=0.023 n=10+10)
Scan1000Limit1_Cockroach-2             497µs ± 2%     497µs ± 3%     ~           (p=0.739 n=10+10)
Scan1000Limit10_Cockroach-2            554µs ± 1%     549µs ± 1%   -1.03%         (p=0.002 n=10+9)
Scan1000Limit100_Cockroach-2          1.04ms ± 1%    1.03ms ± 2%   -0.77%         (p=0.022 n=9+10)
Scan10000FilterLimit1_Cockroach-2      784µs ± 3%     792µs ± 3%     ~           (p=0.247 n=10+10)
Scan10000FilterLimit10_Cockroach-2     916µs ± 3%     900µs ± 3%   -1.80%        (p=0.019 n=10+10)
Scan10000FilterLimit50_Cockroach-2    7.33ms ± 5%    7.23ms ± 1%     ~           (p=0.143 n=10+10)

name                                old alloc/op   new alloc/op   delta
Select1_Cockroach-2                   2.23kB ± 0%    2.23kB ± 0%     ~     (all samples are equal)
Select2_Cockroach-2                    113kB ± 0%     113kB ± 0%     ~             (p=1.000 n=9+9)
Insert1_Cockroach-2                   37.2kB ± 0%    33.0kB ± 0%  -11.19%         (p=0.000 n=10+9)
Insert10_Cockroach-2                  91.5kB ± 0%    92.6kB ± 0%   +1.16%          (p=0.000 n=9+9)
Insert100_Cockroach-2                  629kB ± 0%     626kB ± 0%   -0.51%         (p=0.000 n=10+9)
Insert1000_Cockroach-2                5.55MB ± 1%    5.52MB ± 1%   -0.56%        (p=0.004 n=10+10)
Update1_Cockroach-2                   58.2kB ± 0%    54.1kB ± 0%   -7.17%        (p=0.000 n=10+10)
Update10_Cockroach-2                   158kB ± 0%     159kB ± 0%   +0.71%         (p=0.000 n=9+10)
Update100_Cockroach-2                 1.08MB ± 0%    1.08MB ± 0%   -0.19%         (p=0.000 n=8+10)
Update1000_Cockroach-2                9.21MB ± 1%    9.23MB ± 1%     ~           (p=0.218 n=10+10)
Delete1_Cockroach-2                   43.6kB ± 0%    38.8kB ± 0%  -11.04%         (p=0.000 n=9+10)
Delete10_Cockroach-2                  62.3kB ± 0%    50.9kB ± 0%  -18.35%         (p=0.000 n=9+10)
Delete100_Cockroach-2                  238kB ± 0%     159kB ± 0%  -33.12%         (p=0.000 n=10+9)
Delete1000_Cockroach-2                2.32MB ± 0%    1.53MB ± 0%  -34.16%         (p=0.000 n=10+9)
Scan1_Cockroach-2                     19.3kB ± 0%    19.3kB ± 0%     ~             (p=0.154 n=9+9)
Scan10_Cockroach-2                    26.3kB ± 0%    26.3kB ± 0%     ~            (p=0.700 n=8+10)
Scan100_Cockroach-2                   85.0kB ± 0%    85.0kB ± 0%     ~             (p=0.094 n=9+9)
Scan1000_Cockroach-2                   638kB ± 0%     638kB ± 0%     ~            (p=0.288 n=9+10)
Scan10000_Cockroach-2                 10.6MB ± 0%    10.6MB ± 0%     ~             (p=0.321 n=8+9)
Scan1000Limit1_Cockroach-2            20.1kB ± 0%    20.1kB ± 0%     ~             (p=0.146 n=9+9)
Scan1000Limit10_Cockroach-2           26.9kB ± 0%    26.9kB ± 0%     ~            (p=0.246 n=10+9)
Scan1000Limit100_Cockroach-2          85.3kB ± 0%    85.3kB ± 0%     ~           (p=0.153 n=10+10)
Scan10000FilterLimit1_Cockroach-2     69.6kB ± 0%    69.6kB ± 0%     ~            (p=0.678 n=9+10)
Scan10000FilterLimit10_Cockroach-2    83.2kB ± 0%    83.2kB ± 0%   -0.00%          (p=0.000 n=8+8)
Scan10000FilterLimit50_Cockroach-2     937kB ± 0%     937kB ± 0%   -0.00%          (p=0.015 n=8+8)

name                                old allocs/op  new allocs/op  delta
Select1_Cockroach-2                     48.0 ± 0%      48.0 ± 0%     ~     (all samples are equal)
Select2_Cockroach-2                    2.67k ± 0%     2.67k ± 0%     ~             (p=1.000 n=9+9)
Insert1_Cockroach-2                      473 ± 0%       429 ± 0%   -9.36%         (p=0.000 n=10+8)
Insert10_Cockroach-2                     910 ± 1%       830 ± 0%   -8.75%         (p=0.000 n=10+9)
Insert100_Cockroach-2                  5.11k ± 0%     4.66k ± 0%   -8.84%          (p=0.000 n=9+9)
Insert1000_Cockroach-2                 47.2k ± 0%     43.0k ± 0%   -9.07%        (p=0.000 n=10+10)
Update1_Cockroach-2                      897 ± 0%       853 ± 0%   -4.91%        (p=0.000 n=10+10)
Update10_Cockroach-2                   1.65k ± 0%     1.57k ± 0%   -4.72%         (p=0.000 n=9+10)
Update100_Cockroach-2                  8.71k ± 0%     8.27k ± 0%   -5.07%         (p=0.000 n=8+10)
Update1000_Cockroach-2                 75.1k ± 1%     71.2k ± 0%   -5.14%        (p=0.000 n=10+10)
Delete1_Cockroach-2                      671 ± 0%       610 ± 0%   -9.09%         (p=0.002 n=8+10)
Delete10_Cockroach-2                   1.11k ± 0%     0.89k ± 0%  -19.33%         (p=0.000 n=9+10)
Delete100_Cockroach-2                  5.35k ± 0%     3.61k ± 0%  -32.60%          (p=0.000 n=7+9)
Delete1000_Cockroach-2                 47.7k ± 0%     30.7k ± 0%  -35.76%         (p=0.000 n=10+9)
Scan1_Cockroach-2                        330 ± 0%       330 ± 0%     ~     (all samples are equal)
Scan10_Cockroach-2                       432 ± 0%       432 ± 0%     ~     (all samples are equal)
Scan100_Cockroach-2                    1.34k ± 0%     1.34k ± 0%     ~            (p=0.176 n=10+7)
Scan1000_Cockroach-2                   10.4k ± 0%     10.4k ± 0%     ~            (p=0.137 n=8+10)
Scan10000_Cockroach-2                   101k ± 0%      101k ± 0%     ~            (p=0.328 n=9+10)
Scan1000Limit1_Cockroach-2               345 ± 0%       345 ± 0%     ~     (all samples are equal)
Scan1000Limit10_Cockroach-2              445 ± 0%       445 ± 0%     ~     (all samples are equal)
Scan1000Limit100_Cockroach-2           1.35k ± 0%     1.35k ± 0%     ~     (all samples are equal)
Scan10000FilterLimit1_Cockroach-2        829 ± 0%       829 ± 0%     ~     (all samples are equal)
Scan10000FilterLimit10_Cockroach-2     1.05k ± 0%     1.05k ± 0%     ~     (all samples are equal)
Scan10000FilterLimit50_Cockroach-2     12.8k ± 0%     12.8k ± 0%     ~     (all samples are equal)
```

spencerkimball commented:
@nvanbenschoten was really curious how these now stack up against postgres.

spencerkimball commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[kv/txn_coord_sender_test.go, line 986 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4V_WmV-sVgDOyIyn0:-KF6Nn6Tif8X2fh-EP15:-136839169) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/kv/txn_coord_sender_test.go#L986)):*
I added a comment in `txn_coord_sender.go`

---

*[roachpb/api.proto, line 240 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KEy5hwBcNf8KAMkmGlb-r2-240:-KF6NqxrxxGs8a7ikS-v:-756235959) ([raw file](https://github.com/cockroachdb/cockroach/blob/b206c4c846132b9623cd11fd0cf056763eef5ada/roachpb/api.proto#L240)):*
Yeah, haven't pushed that...was waiting for more comments.

---

*[storage/replica.go, line 1720 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF4Yjn65QY7azVlvDHN:-KF6OWqGFPMbTKPb4iQz:569908873) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa2ac84cec1b5c9a962dd288174fb85a83ea1ccf/storage/replica.go#L1720)):*
Sorry, I understood your suggestion just didn't respond to it directly. It seems even more work to copy the Requests slice, set the `NoopRequests`, then replace the `NoopResponses`. The appends in the code now are ugly but I just tried two different alternatives which were uglier. Feel free to suggest something more concrete though.

---

*[storage/replica.go, line 1694 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF67QT3jW-p6qjVdeXI:-KF6PvxigEr3KHQHSUqI:1378345457) ([raw file](https://github.com/cockroachdb/cockroach/blob/b206c4c846132b9623cd11fd0cf056763eef5ada/storage/replica.go#L1694)):*
One expected error would be write intent error; also definitely uncertainty interval.

Seems not worth the effort and potentially very fragile to white list errors here. I just want the 1PC to happen ONLY if everything is perfect, and otherwise always fall back on the full transactional path.

---

*[storage/replica.go, line 1737 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966#-KF68MV0zv2Mc3hDbPXa:-KF6OPPKBgY00hbEiN_t:-2017001688) ([raw file](https://github.com/cockroachdb/cockroach/blob/b206c4c846132b9623cd11fd0cf056763eef5ada/storage/replica.go#L1737)):*
Yes, it's the most common case for a 1PC txn. The txn arrives, checks timestamp cache and sees a newer read for a key its writing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5966)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
@spencerkimball Take these comparisons with a grain of salt, but here's what I'm seeing when comparing PG numbers to ours (note we are the `new` column). We're actually looking pretty good in the smaller benchmarks.

```
name                      old time/op    new time/op     delta
Select1-2                   73.4µs ± 1%     45.6µs ± 1%      -37.91%  (p=0.000 n=10+10)
Select2-2                    185µs ± 1%     1254µs ± 2%     +578.11%   (p=0.000 n=10+9)
Insert1-2                    772µs ± 6%      669µs ± 1%      -13.30%  (p=0.000 n=10+10)
Insert10-2                   832µs ± 5%      965µs ± 1%      +16.03%  (p=0.000 n=10+10)
Insert100-2                 1.32ms ± 4%     3.57ms ± 3%     +170.98%  (p=0.000 n=10+10)
Insert1000-2                5.44ms ± 4%    31.96ms ± 5%     +487.74%  (p=0.000 n=10+10)
Update1-2                    818µs ± 4%     1127µs ± 2%      +37.70%   (p=0.000 n=10+9)
Update10-2                  1.08ms ± 5%     1.85ms ± 3%      +71.20%  (p=0.000 n=10+10)
Update100-2                 2.06ms ± 3%     7.90ms ± 3%     +283.30%   (p=0.000 n=10+9)
Update1000-2                7.10ms ± 3%    65.93ms ± 3%     +828.45%  (p=0.000 n=10+10)
Delete1-2                    797µs ± 3%      856µs ± 5%       +7.38%   (p=0.000 n=8+10)
Delete10-2                   859µs ± 3%     1185µs ± 2%      +38.04%  (p=0.000 n=10+10)
Delete100-2                 1.14ms ± 7%     5.15ms ± 9%     +351.53%  (p=0.000 n=10+10)
Delete1000-2                3.74ms ± 2%    48.98ms ±12%    +1210.24%  (p=0.000 n=10+10)
Scan1-2                     96.3µs ± 0%    471.4µs ± 6%     +389.35%   (p=0.000 n=9+10)
Scan10-2                     103µs ± 1%      537µs ± 1%     +420.26%  (p=0.000 n=10+10)
Scan100-2                    157µs ± 1%     1000µs ± 2%     +537.77%  (p=0.000 n=10+10)
Scan1000-2                   625µs ± 1%     5475µs ± 3%     +776.41%  (p=0.000 n=10+10)
Scan10000-2                 4.17ms ± 1%    59.92ms ± 4%    +1337.58%  (p=0.000 n=10+10)
Scan1000Limit1-2             108µs ± 1%      493µs ± 2%     +357.30%  (p=0.000 n=10+10)
Scan1000Limit10-2            114µs ± 1%      549µs ± 1%     +381.36%  (p=0.000 n=10+10)
Scan1000Limit100-2           166µs ± 0%     1039µs ± 1%     +525.34%   (p=0.000 n=9+10)
Scan10000FilterLimit1-2      142µs ± 1%      633µs ± 3%     +345.01%  (p=0.000 n=10+10)
Scan10000FilterLimit10-2     413µs ± 1%      746µs ± 3%      +80.44%  (p=0.000 n=10+10)
Scan10000FilterLimit50-2     690µs ± 1%    20455µs ± 3%    +2862.54%  (p=0.000 n=10+10)

name                      old alloc/op   new alloc/op    delta
Select1-2                     440B ± 0%      2232B ± 0%     +407.36%  (p=0.000 n=10+10)
Select2-2                     936B ± 0%    113366B ± 0%   +12011.75%  (p=0.000 n=10+10)
Insert1-2                     272B ± 0%     33030B ± 0%   +12043.22%   (p=0.000 n=10+9)
Insert10-2                    488B ± 0%     92539B ± 0%   +18862.99%  (p=0.000 n=10+10)
Insert100-2                 4.01kB ± 0%   625.34kB ± 0%   +15478.99%    (p=0.000 n=8+9)
Insert1000-2                38.7kB ± 0%   5526.3kB ± 0%   +14177.64%  (p=0.000 n=10+10)
Update1-2                     263B ± 0%     54060B ± 0%   +20455.17%  (p=0.000 n=10+10)
Update10-2                    439B ± 0%    158714B ± 0%   +36053.60%  (p=0.000 n=10+10)
Update100-2                 3.24kB ± 0%  1077.81kB ± 0%   +33134.88%  (p=0.000 n=10+10)
Update1000-2                26.6kB ± 0%   9205.3kB ± 1%   +34548.13%   (p=0.000 n=10+9)
Delete1-2                     240B ± 0%     38768B ± 0%   +16053.49%   (p=0.000 n=10+8)
Delete10-2                    384B ± 0%     50879B ± 0%   +13149.83%   (p=0.000 n=10+9)
Delete100-2                 1.84kB ± 0%   159.01kB ± 0%    +8542.02%  (p=0.000 n=10+10)
Delete1000-2                24.3kB ± 0%   1526.0kB ± 0%    +6186.98%  (p=0.000 n=10+10)
Scan1-2                       456B ± 0%     19279B ± 0%    +4127.83%  (p=0.000 n=10+10)
Scan10-2                      608B ± 0%     26306B ± 0%    +4226.63%   (p=0.000 n=10+9)
Scan100-2                   2.05kB ± 0%    84.97kB ± 0%    +4048.77%   (p=0.000 n=10+8)
Scan1000-2                  16.5kB ± 0%    638.2kB ± 0%    +3776.78%   (p=0.000 n=9+10)
Scan10000-2                  161kB ± 0%    10564kB ± 0%    +6479.72%   (p=0.000 n=10+7)
Scan1000Limit1-2              503B ± 0%     20083B ± 0%    +3892.69%   (p=0.000 n=10+9)
Scan1000Limit10-2             656B ± 0%     26943B ± 0%    +4007.13%   (p=0.000 n=10+9)
Scan1000Limit100-2          2.10kB ± 0%    85.32kB ± 0%    +3970.53%  (p=0.000 n=10+10)
Scan10000FilterLimit1-2       599B ± 0%     36471B ± 0%    +5988.61%  (p=0.000 n=10+10)
Scan10000FilterLimit10-2      895B ± 0%     49815B ± 0%    +5465.92%  (p=0.000 n=10+10)
Scan10000FilterLimit50-2    2.17kB ± 0%  2382.91kB ± 0%  +109459.09%  (p=0.000 n=10+10)

name                      old allocs/op  new allocs/op   delta
Select1-2                     14.0 ± 0%       48.0 ± 0%     +242.86%  (p=0.000 n=10+10)
Select2-2                     23.0 ± 0%     2670.0 ± 0%   +11508.70%   (p=0.000 n=10+8)
Insert1-2                     9.00 ± 0%     429.00 ± 0%    +4666.67%   (p=0.000 n=10+8)
Insert10-2                    18.0 ± 0%      830.0 ± 0%    +4511.11%   (p=0.000 n=10+7)
Insert100-2                    109 ± 0%       4659 ± 0%    +4174.31%   (p=0.000 n=10+9)
Insert1000-2                 1.01k ± 0%     42.96k ± 0%    +4157.43%   (p=0.000 n=10+9)
Update1-2                     9.00 ± 0%     853.00 ± 0%    +9377.78%  (p=0.000 n=10+10)
Update10-2                    18.0 ± 0%     1573.6 ± 0%    +8642.22%  (p=0.000 n=10+10)
Update100-2                    109 ± 0%       8273 ± 0%    +7489.82%  (p=0.000 n=10+10)
Update1000-2                 1.01k ± 0%     71.13k ± 0%    +6949.14%   (p=0.000 n=10+9)
Delete1-2                     9.00 ± 0%     610.00 ± 0%    +6677.78%   (p=0.000 n=10+9)
Delete10-2                    18.0 ± 0%      894.0 ± 0%    +4866.67%   (p=0.000 n=10+7)
Delete100-2                    108 ± 0%       3608 ± 0%    +3240.37%  (p=0.000 n=10+10)
Delete1000-2                 1.01k ± 0%     30.67k ± 0%    +2939.60%   (p=0.000 n=10+9)
Scan1-2                       17.0 ± 0%      330.0 ± 0%    +1841.18%  (p=0.000 n=10+10)
Scan10-2                      35.0 ± 0%      432.0 ± 0%    +1134.29%   (p=0.000 n=10+9)
Scan100-2                      215 ± 0%       1342 ± 0%     +524.00%  (p=0.000 n=10+10)
Scan1000-2                   2.02k ± 0%     10.37k ± 0%     +414.58%  (p=0.000 n=10+10)
Scan10000-2                  20.0k ± 0%     100.8k ± 0%     +403.49%   (p=0.000 n=10+8)
Scan1000Limit1-2              17.0 ± 0%      345.0 ± 0%    +1929.41%  (p=0.000 n=10+10)
Scan1000Limit10-2             36.0 ± 0%      445.0 ± 0%    +1136.11%  (p=0.000 n=10+10)
Scan1000Limit100-2             216 ± 0%       1354 ± 0%     +526.85%  (p=0.000 n=10+10)
Scan10000FilterLimit1-2       20.0 ± 0%      629.0 ± 0%    +3045.00%  (p=0.000 n=10+10)
Scan10000FilterLimit10-2      56.0 ± 0%      855.0 ± 0%    +1426.79%  (p=0.000 n=10+10)
Scan10000FilterLimit50-2       216 ± 0%      35721 ± 0%   +16437.36%  (p=0.000 n=10+10)
```

spencerkimball commented:
@nvanbenschoten Do the allocs even make any sense in this context? The CRDB ones are measuring full server allocs, not just client ones, right?

nvanbenschoten commented:
Yeah good point, the allocs measured are completely misleading. Removed from the post.

bdarnell commented:
This PR slipped by me, but looking at it now it scares me from a migration perspective because it changes things downstream of raft and is going to affect a lot of commands. I *think* it's OK because of the auto-gc of local intents and transaction records (except for esoteric cases: if the `BeginTransaction` or `EndTransaction` call specified a key in their `Header` that didn't match the `Transaction`, it would fail before but now it will pass). @spencerkimball and @tschottdorf are you confident that this produces byte-for-byte identical results to the old code or do we need to revert this change until we can have a proper migration for it?

I also have a concern around merges: could a transaction be attempted in a distributed mode on the first try, then retried as a 1PC transaction because the range merged underneath it? In that case we'd still need to process the on-disk transaction record from the first attempt. I'm not sure whether the DistSender/TxnCoordSender stack would allow that or not.

tschottdorf commented:
I see a slight chance for merges to diverge (but luckily we do not have them), though I can't come up with a concrete scenario. If a first execution of the Txn lays down intents and then retries in 1PC fashion, the servers which use the optimization not recognize their intents, so they'll run the command without the optimization.

As for whether executing the MVCC ops without the transaction can make a difference - that's a really good question and we ought to have more testing for that. It only matters in cases in which not having the transaction succeeds (since otherwise we fall back to with-txn execution anyway), but succeeds in a different way than having the transaction. That rules out all intent-related issues (since running into an intent at that level always means an error), WriteTooOld, uncertainty restarts and pretty much everything I can think of. But can I guarantee that none of the code paths have possibly changed? Certainly not. That's a general problem with migrations - it's impossible to traverse the space of all possible chains of events which could lead to any inconsistency. The only way in which I could see that happen is if the program and some TLA+ spec were really tightly coupled.

In any case, the conditions for the optimization should be tightened to only accept the initial sequence number and epoch to put an end to any retry-related shenanigans.

I personally wouldn't pull this change (it's a big win and the migration is going to be horrible no matter how much infrastructure we build) in the absence of serious users and us not supporting mixed clusters, but in a production setting I would (though with that mindset we could basically freeze development for a month at least).

Curious whether @spencerkimball has anything to add.

spencerkimball commented:
I would add that I'm becoming convinced that the suggestion we pull command execution up above raft is going to be necessary. It's simply not [practically] possible to reason about replica divergence given the branching of execution paths we already have in the code base. Going forward, we won't be able to make non-trivial changes and that's not acceptable.

tschottdorf commented:
It's hard to disagree with that argument. Let's continue that discussion in
https://github.com/cockroachdb/cockroach/pull/5985 (which is where you
initially proposed that idea).

On Thu, Apr 14, 2016 at 8:15 AM Spencer Kimball <notifications@github.com>
wrote:

> I would add that I'm becoming convinced that the suggestion we pull
> command execution up above raft is going to be necessary. It's simply not
> [practically] possible to reason about replica divergence given the
> branching of execution paths we already have in the code base. Going
> forward, we won't be able to make non-trivial changes and that's not
> acceptable.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/5966#issuecomment-209910456>
>
-- 

-- Tobias
